File: algo_2021/w2.-two-pointer/problems/415.valid_palindrom/415_valid_palindrom.java
public class Solution {
    /**
     * @param s: A string
     * @return: Whether the string is a valid palindrome
     */
    public boolean isPalindrome(String s) {
        // write your code here
        
        // two pointers
        int left = 0, right = s.length() - 1;
        while (left < right) {
            
            // if non-alphanumeric characters, keep traversing
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }
            while(left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }
            // if not the same, break earlier
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
             // keep traversing
            left++;
            right--;
            }
         return true;
    }
}

File: algo_2021/w2.-two-pointer/415.valid_palindrom/415_valid_palindrom.py
class Solution:
    """
    @param s: A string
    @return: Whether the string is a valid palindrome
    """
    def isPalindrome(self, s):
        # write your code here
        
        # two pointers
        left, right = 0, len(s) - 1
        
        while left < right:
            # if non-alphanumeric characters, keep traversing
            while left < right and not s[left].isalnum():
                left+=1
            while left < right and not s[right].isalnum():
                right-=1
            # if not the same, break earlier
            if s[left].lower() != s[right].lower():
                return False
            # keep traversing
            left+=1
            right-=1
        return True